mitodl/mitxonline
# Updates discounts in the staff dashboard to reflect the current state of the art

#### Pre-Flight checklist


#### What are the relevant tickets?

Closes #1266

#### What's this PR do?

Updates the Discounts section of the staff dashboard app to be more in line with the current state of the art for the system:
- Dashboard
   - Adds search functionality
   - Updates the Flexible Price Tier tag to read Financial Assistance Tier
   - Adds the activation and expiration dates
   - Adds a Redeemed tag if the discount is single-use and has been used
- Create/Edit
   - Refactored out the form into a separate reusable component
   - Added maximum redemption field (hidden unless the discount is unlimited use)
   - Added fields for activation and expiry
   - Added financial assistance tier field (defaults to not)
- Edit
   - Adds tables to manage assigned users and products for the discount
- Show
   - Added additional fields to the top, in a two-column format: max redemptions (if applicable) and activation/expiration dates  
   - Added the Redeemed tag for single-use discounts that have been redeemed
   - Added a table to show the tiers the discount is used for (if it's a financial assistance discount)
   - Moved the Redemptions table up just under the discount information
- Meta
   - Fixed the route for `/staff-dashboard` so that deep links into the dashboard will actually work now 
   - Added a serializer for the financial assistance tiers
   - Added endpoints for tiers for a given discount 

#### How should this be manually tested?

Automated tests should pass. 

Discount Dashboard:
- By default, the list should show all discounts in the system with working pagination controls.
- Active From and Expires should be set if the discount is time-limited. 
- Discounts should be tagged "Financial Assistance Tier" if the discount is for financial assistance.
- Discounts should be tagged "Redeemed" if they're one-time or one-time-per-user and the discount has been used.
- Search should work, and search items should be ANDed.

Create:
- Should work.
- The form's Financial Assistance field should default to Regular Discount.
- The activation and expiration date fields should default to null.
- If the discount is set to Unlimited, you should be able to specify a Max Redemptions value.
- You should not see a Products or User Assignments table.

Edit:
- Should work.
- The Financial Assistance, Activation Date, and Expiration Date fields should all populate properly. 
- Max Redemptions should appear if the discount is Unlimited use and should populate properly.
- You should see the Products table.
   - Products should display any products the discount has been assigned to.
   - Clicking Delete should remove the product from the discount. 
   - Clicking Add Product should open a modal with a search field.
   - You should be able to search for products based on product name or readable ID. (Note that most products use the courseware readable ID as the product name so you may want to add some products that don't follow that convention to test with.) 
   - Clicking Add should add products to the list, and they should show up in the underlying table. (You should be able to see this if the Products table isn't obscured.) 
   - Clicking Close should close the modal.
- You should see the Assigned Users table.
   - The table should show any users assigned to the discount.
   - Operations should work in the same manner as the Products table.
   - The Assign Users modal should include some text describing how the user discounts work. 

View:
- The discount should display the Redeemed tag if the discount is one-time or one-time-per-user and has been used.
- If the discount has an activation or expiration date, it should be displayed on the right.
- If the discount has a usage limit, it should be displayed on the right.  
- The Redemptions, Assigned Users, and Products tables should all work as expected.
- Assigned Users and Products should _not_ display the Add Products or Assign Users buttons. 
- The Financial Assistance Tiers table should be visible and should show any tiers that are using the discount. 
- The tables should be in the order of Redemptions, Financial Assistance Tiers, Assigned Users, Products. 

To test the URL change: the easiest way to do this is just to go to a page within the staff dashboard (any of them) and hit reload. The route for the staff dashboard only matched on `^/staff-dashboard` so that would fail if you were going to a specific resource within it (for example, `/staff-dashboard/discounts`) - the new regex should fix the issue.


#### Screenshots (if appropriate)

Dashboard: 
![image](https://user-images.githubusercontent.com/945611/210883322-3b9427fa-c7d9-4b6c-b50b-5fc634c2f9ac.png)

Create:
![image](https://user-images.githubusercontent.com/945611/211878151-6174cf26-acd5-4fa5-a486-967784ae3d7f.png)


Edit:
![image](https://user-images.githubusercontent.com/945611/211877837-f3df47ec-16a7-446e-9c87-1d9776dff373.png)

Edit - Assign Products dialog (with search results):
![image](https://user-images.githubusercontent.com/945611/211877944-c65e7425-8536-4caf-ad15-f676977f477f.png)

Edit - Assign Users dialog (with search results):
![image](https://user-images.githubusercontent.com/945611/211878024-62840bcf-766c-4f11-bf75-c065156b6998.png)

Show:
![image](https://user-images.githubusercontent.com/945611/210883552-aca96820-f5e9-441d-87c2-c3789f796e36.png)

Show (w/ redemptions, assigned users, assigned products):
![Screenshot 2023-01-05 at 3 52 25 PM](https://user-images.githubusercontent.com/945611/210887080-6770d512-07cb-4aa4-8322-7aad152feb7d.png)
